### DISCUSSION TITLE
Thrust PTX version check

### DISCUSSION THREAD

**Discussion by NamjaeChoi** [NVIDIA/cccl#2005] [Thrust] [answered]
Hello.

We want to use a code compiled with a lower CC (7.0) on a higher CC (8.9) GPU. This is because we have multiple HPC systems with different GPUs, and we want to build a container that is compatible for all our HPC systems. But Thrust complains that the code was not compiled with a proper CC if the major version does not match. What is the reason for this limitation and can this be lifted?

Thank you.

**Comment by jrhemstad** [comment]
Can you provide an example of what you're seeing? Is this a compile time error? Runtime? A reproducer would go a long way!

**Reply by NamjaeChoi** [reply]
It's a runtime error:

```
This program was not compiled for SM 89  
: cudaErrorInvalidDevice: invalid device ordinal
```

and is coming from `get_ptx_version()`.

**Reply by NamjaeChoi** [reply]
Some backtraces:
```
#5  0x00007fffe86cfccd in thrust::cuda_cub::throw_on_error (status=cudaErrorInvalidDevice, 
    msg=0x7fffffffa7a0 "This program was not compiled for SM 89  \n")
    at /usr/local/cuda/include/thrust/system/cuda/detail/util.h:282
#6  0x00007fffe86d000f in thrust::cuda_cub::core::get_ptx_version ()
    at /usr/local/cuda/include/thrust/system/cuda/detail/core/util.h:666
#7  0x00007fffea321354 in thrust::cuda_cub::core::AgentLauncher<thrust::cuda_cub::__parallel_for::ParallelForAgent<thrust::cuda_cub::__uninitialized_fill::functor<thrust::device_ptr<int>, int>, unsigned long> >::get_plan (d_ptr=0x0)
    at /usr/local/cuda/include/thrust/system/cuda/detail/core/agent_launcher.h:492
#8  0x00007fffea31f213 in thrust::cuda_cub::__parallel_for::parallel_for<thrust::cuda_cub::__uninitialized_fill::functor<thrust::device_ptr<int>, int>, unsigned long> (num_items=181543, f=..., 
    stream=0x1) at /usr/local/cuda/include/thrust/system/cuda/detail/parallel_for.h:136
#9  0x00007fffea31ce93 in thrust::cuda_cub::parallel_for<thrust::cuda_cub::tag, thrust::cuda_cub::__uninitialized_fill::functor<thrust::device_ptr<int>, int>, unsigned long> (policy=..., f=..., 
    count=181543) at /usr/local/cuda/include/thrust/system/cuda/detail/parallel_for.h:161
#10 0x00007fffea31a75c in thrust::cuda_cub::uninitialized_fill_n<thrust::cuda_cub::tag, thrust::device_ptr<int>, unsigned long, int> (policy=..., first=..., count=181543, x=@0x7fffffffaa18: 0)
    at /usr/local/cuda/include/thrust/system/cuda/detail/uninitialized_fill.h:84
#11 0x00007fffea317c36 in thrust::uninitialized_fill_n<thrust::cuda_cub::tag, thrust::device_ptr<int>, unsigned long, int> (exec=..., first=..., n=181543, x=@0x7fffffffaa18: 0)
    at /usr/local/cuda/include/thrust/detail/uninitialized_fill.inl:52
#12 0x00007fffea31778f in thrust::detail::allocator_traits_detail::default_construct_range<thrust::device_allocator<int>, thrust::device_ptr<int>, unsigned long> (a=..., p=..., n=181543)
    at /usr/local/cuda/include/thrust/detail/allocator/default_construct_range.inl:95
#13 0x00007fffea315042 in thrust::detail::default_construct_range<thrust::device_allocator<int>, thrust::device_ptr<int>, unsigned long> (a=..., p=..., n=181543)
    at /usr/local/cuda/include/thrust/detail/allocator/default_construct_range.inl:106
#14 0x00007fffea31251b in thrust::detail::contiguous_storage<int, thrust::device_allocator<int> >::default_construct_n (this=0xfc1e500, first=..., n=181543)
    at /usr/local/cuda/include/thrust/detail/contiguous_storage.inl:256
#15 0x00007fffea3105a4 in thrust::detail::vector_base<int, thrust::device_allocator<int> >::default_init (this=0xfc1e500, n=181543) at /usr/local/cuda/include/thrust/detail/vector_base.inl:245
#16 0x00007fffea30ee5b in thrust::detail::vector_base<int, thrust::device_allocator<int> >::vector_base (this=0xfc1e500, n=181543) at /usr/local/cuda/include/thrust/detail/vector_base.inl:62
#17 0x00007fffea30d3b7 in thrust::device_vector<int, thrust::device_allocator<int> >::device_vector (this=0xfc1e500, n=181543) at /usr/local/cuda/include/thrust/device_vector.h:87
```

**Reply by alliepiper** [reply]
This will depend on how the executable is compiled, in particular the options listed here:

https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#options-for-steering-gpu-code-generation

It is possible to compile a suitable binary using those flags.

**Reply by alliepiper** [reply]
And if you're using CMake, these flags can be set using the `CMAKE_CUDA_ARCHITECTURES` option:

https://cmake.org/cmake/help/latest/variable/CMAKE_CUDA_ARCHITECTURES.html

**Reply by NamjaeChoi** [reply]
We can set the architecture specific to the GPU we are running, but that is not what we prefer. We would like to set the architecture to the lowest common version of our GPUs so that it can be run on all our GPUs. Setting multiple architectures is also limited due to one of our dependencies (Kokkos).

**Reply by jrhemstad** [reply, marked as answer]
The error you're seeing isn't from any requirements Thrust is enforcing. This is just propagating an error from a CUDA runtime API call. 

Can you show exactly how you're specifying the architecture flags for your build? My guess is it's only embedding SASS and not PTX. Forward compatibility only works when PTX is embedded in the binary.

**Reply by NamjaeChoi** [reply]
It's an external package (PETSc) that supplies the flags, but it has a format `-gencode arch=compute_xx,code=sm_xx`.

**Reply by pauleonix** [reply]
@NamjaeChoi In case you are not aware of that: You can specify multiple architectures s.t. multiple versions of the compiled kernels are included in the library/executable. See [Fatbinaries](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#fatbinaries).

**Reply by NamjaeChoi** [reply]
@jrhemstad Thank you, your diagnosis was correct. Based on a StackOverflow thread (https://stackoverflow.com/questions/26232500/does-code-sm-x-embed-only-binary-cubin-code-or-also-ptx-code-or-both), we changed it to `-arch=sm_xx`, and it now runs okay.
